### PR TITLE
libjobspec and job-ingest intial work

### PR DESCRIPTION
**Pull request by morrone**
Here's the current state of libjobspec and job-ingest.

### CODE REVIEW

**Comment by dongahn** [comment]
Thanks you @morrone for submitting this PR. I understand it is initial work, but this should be very useful for me to be furthering the resource service piece. I will add some inline comments if I have any feedback.

File: src/common/libjobspec/jobspec.cpp [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+    count.oper = cnode["operator"].as<char>();
+    switch (count.oper) {
+    case '+':
+    case '-':
+    case '^':
+        break;
+    default:
+        throw invalid_argument("Invalid count operator");
+    }
+
```
Comment (dongahn) [inline src/common/libjobspec/jobspec.cpp]: Any reason not to add some validation for the operand  field against the operator? Seems -1 would be an error if the operand is +, 2 would be an error if the operand is -. In general, more validation the less work at the client side...

Oh BTW, what would be the value of max if only min is given?

File: src/common/libflux/msg_handler.h [hunk 2]
```diff
@@ -27,7 +27,7 @@ void flux_msg_handler_deny_rolemask (flux_msg_handler_t *w, uint32_t rolemask);
 
 struct flux_msg_handler_spec {
     int typemask;
-    char *topic_glob;
+    const char *topic_glob;
```
Comment (garlick) [inline src/common/libflux/msg_handler.h]: Maybe this was pulled in by mistake?  If intended, it should probably be split out to another commit.

File: src/modules/job-ingest/job-ingest.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <flux/jobspec.hpp>
+
+using namespace Flux::Jobspec;
+
+extern "C" {
+#include <flux/core.h>
```
Comment (dongahn) [inline src/modules/job-ingest/job-ingest.cpp]: This is not for your code @morrone but since we will increaingly use C++, it seems it would be useful to add __cplusplus support into flux header files.

**Comment by garlick** [comment]
> This is not for your code @morrone but since we will increaingly use C++, it seems it would be useful to add __cplusplus support into flux header files.

That came up before and I can't remember where we left it, why it isn't done already.  Perhaps one of the C++ aficionados should open a flux-core issue on that one.

File: src/common/libjobspec/jobspec.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the terms and conditions of the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
+ *  See also:  http://www.gnu.org/licenses/
+\*****************************************************************************/
+
+#ifndef jobspec_hpp
```
[the remark this answers is not in the document]
Comment (morrone) [reply src/common/libjobspec/jobspec.hpp]: Nope, I can do that.

File: src/common/libjobspec/jobspec.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    unsigned int version;
+    std::vector<Resource> resources;
+    std::vector<Task> tasks;
+    std::unordered_map<std::string,
+        std::unordered_map<std::string, std::string>> attributes;
+    Jobspec() = default;
+    Jobspec(const YAML::Node&);
+    Jobspec(std::istream &is): Jobspec{YAML::Load(is)} {}
+    Jobspec(std::string &s): Jobspec{YAML::Load(s)} {}
+};
```
[the remark this answers is not in the document]
Comment (morrone) [reply src/common/libjobspec/jobspec.hpp]: Yeah, error handling isn't nailed down yet.  I'd like to make custom exception with position information about where the error occurred.  But I took a pause on that work when I downgraded from yaml-cpp 0.5.3 to 0.5.1 (which is in EPEL).  That version lost the line number information.  Whatever I do will definitely derive from std:exception.  Maybe I can still a placeholder exception in there for now.

File: src/common/libjobspec/jobspec.cpp (hunk 1 above)
Comment (morrone) [reply src/common/libjobspec/jobspec.cpp]: Sure, if someone want to decide what hte operators are going to be, I can add more validation.

Comment (morrone) [reply src/common/libjobspec/jobspec.cpp]: Oh, and at the moment all four fields are required.  The idea to allowed defaults in the count area is not yet in the RFC, so it doesn't appear in the code yet either.

File: src/common/libflux/msg_handler.h (hunk 2 above)
Comment (morrone) [reply src/common/libflux/msg_handler.h]: Actually, it was intentional. It was required to allow the header to compiler under the C++ compiler.  I   should note that in the comment.

File: src/modules/job-ingest/job-ingest.cpp (hunk 3 above)
Comment (morrone) [reply src/modules/job-ingest/job-ingest.cpp]: Yes!  Actually, that is exactly what I tried to do.  I went through and made the change to all the C headers in flux core.  Unfortunately, our python interface auto-generator threw a fit when it hit those statements.  It seemed like it wasn't properly preprocessing the files and things were still in its file that should have been removed at preprocessing time.   It looked like it was going to take a while to figure out how to make it happy, so I backtracked and took the easy route.

But yes, that would be the better approach in the long run.

**Comment by garlick** [comment]
The submit command and ingest module should probably be dropped from this PR IMHO, as topically they are distinct, and they don't seem to be very far along compared to the jobspec library and validator.

This PR should also add the yaml prereq to travis so test coverage is measured.  Would it be fairly simple to have sharness push the jobspec RFC examples through the validator to get a start on coverage?

File: configure.ac [hunk 6]
```diff
@@ -313,6 +321,9 @@ AC_CONFIG_LINKS([ \
   t/fluxometer.lua:t/fluxometer.lua \
 ])
 
+AC_CONFIG_FILES([etc/rc1], [chmod +x etc/rc1])
+AC_CONFIG_FILES([etc/rc3], [chmod +x etc/rc3])
+
 AC_OUTPUT
```
Comment (grondo) [inline configure.ac]: Optional, but it would be nice if fixes like this were split out from the jobspec related changes.

File: etc/.gitignore [hunk 7]
```diff
@@ -1 +1,4 @@
 /flux
+/flux.service
+/rc1
+/rc3
```
Comment (grondo) [inline etc/.gitignore]: Thanks for fixing this, but as before it would be nice (not mandatory though) if the `.gitignore` changes unrelated to jobspec were under a separate commit. (just my .02)

File: src/common/libjobspec/Makefile.am [hunk 8]
```diff
@@ -0,0 +1,13 @@
+if ENABLE_JOBSPEC
```
[the remark this answers is not in the document]
Comment (morrone) [reply src/common/libjobspec/Makefile.am]: I was wondering about that.  I believe those can be set globally.  But maybe there are things to which we don't want to apply those flags?

File: src/modules/job-ingest/Makefile.am [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# AM_CFLAGS = \
+# 	@GCCWARN@ \
+# 	$(CODE_COVERAGE_CFLAGS)
+
+# AM_LDFLAGS = \
+# 	$(CODE_COVERAGE_LDFLAGS)
+
```
[the remark this answers is not in the document]
Comment (morrone) [reply src/modules/job-ingest/Makefile.am]: Yeah, those flags were for C, so I didn't want to assume they were right for C++.  If nothing else, they probably need to be added to things like AM_CXXFLAGS or something.  You are correct, I need to actually look at them now and see what to apply for CXX.

File: configure.ac (hunk 6 above)
Comment (morrone) [reply configure.ac]: They are in a separate commit.  I'll just drop the job ingest stuff from this PR like Jim suggested.

**Comment by morrone** [comment]
This pull request is replaced with pull request #1201.
